torch/_inductor/codegen/cpp.py

```diff
@@ -666,6 +666,14 @@ def index_expr(expr, dtype):
 class CppOverrides(OpOverrides):
     """Map element-wise ops to C++"""
 
+    @staticmethod
+    def add(a, b):
+        return f"decltype({a})({a} + {b})"
+
+    @staticmethod
+    def sub(a, b):
+        return f"decltype({a})({a} - {b})"
+
     @staticmethod
     def mul(a, b):
         return f"decltype({a})({a} * {b})"
```

Comment:
Type-casting with `decltype({a})` doesn't seem right. A better approach is to use the dtype from type propagation which should consider both `a` and `b`, their dtypes and whether they are scalar or not, following the type promotion rules of pytorch.